app/src/main/java/com/keylesspalace/tusky/ViewMediaActivity.kt

```diff
@@ -0,0 +1,249 @@
+            val id = item.itemId
+            when (id) {
+                R.id.action_download -> downloadImage()
+                R.id.action_open_status -> onOpenStatus()
+                R.id.action_share_media -> shareImage()
+            }
+            true
+        }
+
+        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LOW_PROFILE
+        window.statusBarColor = Color.BLACK
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        if(attachments != null) {
+            menuInflater.inflate(R.menu.view_media_toolbar, menu)
+            return true
+        }
+        return false
+    }
+
+    override fun onBringUp() {
+        supportStartPostponedEnterTransition()
+    }
+
+    override fun onDismiss() {
+        supportFinishAfterTransition()
+    }
+
+    override fun onPhotoTap() {
+        toolbarVisible = !toolbarVisible
+        for (listener in toolbarVisibilityListeners) {
+            listener.onToolbarVisiblityChanged(toolbarVisible)
+        }
+        val visibility = if(toolbarVisible){ View.VISIBLE } else { View.INVISIBLE }
+        val alpha = if(toolbarVisible){ 1.0f } else { 0.0f }
+
+        toolbar.animate().alpha(alpha)
+                .setListener(object: AnimatorListenerAdapter() {
+                    override fun onAnimationEnd(animation: Animator) {
+                        toolbar.visibility = visibility
+                        animation.removeListener(this)
+                    }
+                })
+                .start()
+    }
+
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
+        when (requestCode) {
+            PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE -> {
+                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    downloadImage()
+                } else {
+                    doErrorDialog(toolbar, R.string.error_media_download_permission, R.string.action_retry) { _ -> downloadImage() }
+                }
+            }
+        }
+    }
+
+    private fun downloadImage() {
+        downloadFile(attachments!![viewPager.currentItem].attachment.url)
+    }
+
+    private fun onOpenStatus() {
+        val attach = attachments!![viewPager.currentItem]
+        startActivityWithSlideInAnimation(ViewThreadActivity.startIntent(this, attach.statusId, attach.statusUrl))
+    }
+
+    private fun shareImage() {
+        val directory = applicationContext.getExternalFilesDir("Tusky")
+        if (directory == null || !(directory.exists())) {
+            Log.e(TAG, "Error obtaining directory to save temporary media.")
+            return
+        }
+
+        val attachment = attachments!![viewPager.currentItem].attachment
+        val context = applicationContext
+        val file = File(directory, MediaUtils.getTemporaryMediaFilename("png"))
+
+        Picasso.with(context).load(Uri.parse(attachment.url)).into(object: Target {
```

Comment:
We could, probably, avoid keeping image in memory if we didn't use Picasso but we would lose caching :thinking: